File: apps/grokking-ml-fsharp-lib/src/GrokkingMlLib/Nb07NaiveBayes.fs
/// 原著ノートブック #07 `Chapter_08_Naive_Bayes/Coding_naive_Bayes.ipynb`。
///
/// メール 5728 通からスパム判定器をナイーブベイズで作る。
/// **この回だけ原著が scikit-learn を使わず、pandas と辞書だけで書いている。**
/// だからライブラリの差が出ず、F# でも原著の数値をそのまま再現できる。
///
/// 原著は最後の確率計算で `np.compat.long` を使っており、これが浮動小数点数を
/// **整数に切り捨てる**。切り捨てないと原著の出力と 8 桁目から食い違うので、
/// その挙動もそのまま写してある。
module GrokkingMlLib.Nb07NaiveBayes

open System
open System.Collections.Generic
open System.Numerics
open System.Text.RegularExpressions
open Deedle

/// ある単語が、スパムとハムそれぞれ何通に現れたか。
///
/// 原著は 1 から数え始める。ラプラス平滑化にあたり、
/// 「一度も見ていない側」の確率が 0 になるのを防ぐ。
type WordCounts = { Spam: int; Ham: int }

/// 数え始めの値
let initialCounts = { Spam = 1; Ham = 1 }

/// 学習に使ったメール全体の統計
type Corpus =
    { Total: int; Spam: int }

    member this.Ham = this.Total - this.Spam

    /// 事前確率。何も情報が無いときにスパムと判断する確率
    member this.SpamProbability = float this.Spam / float this.Total

/// 単語ごとの出現数と、コーパス全体の統計
type NaiveBayesModel =
    { Words: IReadOnlyDictionary<string, WordCounts>
      Corpus: Corpus }

/// メールと、スパムかどうかのラベル
type Emails = { Texts: string[]; Spam: int[] }

let private whitespace = Regex(@"\s+", RegexOptions.Compiled)

/// メール本文を、重複を除いた小文字の単語集合にする。
///
/// 原著は `list(set(text.lower().split()))`。**同じ単語が何度出ても 1 回** と
/// 数えるのがこの実装の前提である。
///
/// Python の `split()` は空白の連続をまとめて区切る。`Regex.Split` は
/// 空文字列を残すので、そこを落とす必要がある。
let processEmail (text: string) =
    whitespace.Split(text.ToLowerInvariant())
    |> Array.filter (fun token -> token <> "")
    |> Set.ofArray

/// メールのデータセットを読み込む。5728 通、うち 1368 通がスパム。
///
/// このファイルは約 8.5 MB あるためリポジトリには含めていない。
let loadEmails () =
    let frame = Datasets.loadFrame "emails.csv"

    { Texts = frame.GetColumn<string>("text") |> Series.values |> Array.ofSeq
      Spam = frame.GetColumn<int>("spam") |> Series.values |> Array.ofSeq }

/// 単語ごとに、スパム・ハムそれぞれの出現通数を数える
let train (emails: Emails) =
    let words = Dictionary<string, WordCounts>()

    for i in 0 .. emails.Texts.Length - 1 do
        for word in processEmail emails.Texts.[i] do
            let current =
                match words.TryGetValue word with
                | true, counts -> counts
                | _ -> initialCounts

            words.[word] <-
                if emails.Spam.[i] = 1 then
                    { current with Spam = current.Spam + 1 }
                else
                    { current with Ham = current.Ham + 1 }

    { Words = words
      Corpus =
        { Total = emails.Texts.Length
          Spam = emails.Spam |> Array.sumBy (fun s -> if s = 1 then 1 else 0) } }

/// 単語 1 つだけを見たときの、スパムである確率。
///
/// ベイズの定理そのものではなく、単純に「その単語を含むメールのうち
/// スパムの割合」を返す。原著もそう書いている。
let predictBayes (model: NaiveBayesModel) (word: string) =
    let counts = model.Words.[word.ToLowerInvariant()]
    float counts.Spam / float (counts.Spam + counts.Ham)

/// メール全体を見たときの、スパムである確率。
///
/// 「単語の出現が互いに独立」と仮定して、単語ごとの尤度比を掛け合わせる。
/// 語彙にない単語は無視するので、知らない単語ばかりのメールは事前確率に落ちる。
///
/// 最後に整数へ切り捨てているのは、原著の `np.compat.long` に合わせるため。
/// 切り捨てないと 8 桁目から数値が変わる。
let predictNaiveBayes (model: NaiveBayesModel) (email: string) =
    let corpus = model.Corpus

    let productSpams, productHams =
        processEmail email
        |> Set.fold
            (fun (spams, hams) word ->
                match model.Words.TryGetValue word with
                | true, counts ->
                    spams * (float counts.Spam / float corpus.Spam * float corpus.Total),
                    hams * (float counts.Ham / float corpus.Ham * float corpus.Total)
                | _ -> spams, hams)
            (float corpus.Spam, float corpus.Ham)

    // 原著の np.compat.long と同じ切り捨て。値は 10^20 を超えることがあるので
    // int64 では溢れる。Python の int は多倍長なので BigInteger で受ける
    let truncatedSpams = BigInteger productSpams
    let truncatedHams = BigInteger productHams

    // Python は int / int を多倍長のまま計算してから float にする。
    // decimal では桁が足りないので、比を BigInteger の割り算で作ってから float にする
    let scale = BigInteger.Pow(BigInteger 10, 20)
    float (truncatedSpams * scale / (truncatedSpams + truncatedHams)) / 1e20

File: apps/grokking-ml-python-lib/src/grokking_ml_lib/nb07_naive_bayes.py
"""原著ノートブック #07 `Chapter_08_Naive_Bayes/Coding_naive_Bayes.ipynb`。

メール 5728 通からスパム判定器をナイーブベイズで作る。
**この回だけ scikit-learn を使わず、pandas と辞書だけで書かれている。**
だからライブラリの差が出にくく、3 言語とも原著の数値をそのまま再現できる。

原著は最後の確率計算で `np.compat.long` を使っており、これが浮動小数点数を
**整数に切り捨てる**。切り捨てないと原著の出力と 8 桁目から食い違うので、
その挙動もそのまま写してある。
"""

from __future__ import annotations

from dataclasses import dataclass

import pandas as pd

from grokking_ml_lib.datasets import dataset_path


@dataclass
class WordCounts:
    """ある単語が、スパムとハムそれぞれ何通に現れたか。

    原著は 1 から数え始める。ラプラス平滑化にあたり、
    「一度も見ていない側」の確率が 0 になるのを防ぐ。
    """

    spam: int = 1
    ham: int = 1


@dataclass
class Corpus:
    """学習に使ったメール全体の統計。"""

    total: int
    spam: int

    @property
    def ham(self) -> int:
        return self.total - self.spam

    @property
    def spam_probability(self) -> float:
        """事前確率。何も情報が無いときにスパムと判断する確率。"""
        return self.spam / self.total


@dataclass
class NaiveBayesModel:
    """単語ごとの出現数と、コーパス全体の統計。"""

    words: dict[str, WordCounts]
    corpus: Corpus


def load_emails() -> pd.DataFrame:
    """メールのデータセットを読み込む。5728 通、うち 1368 通がスパム。

    このファイルは約 8.5 MB あるためリポジトリには含めていない。
    未取得なら `dataset_path` が原著リポジトリから取得する。
    """
    return pd.read_csv(dataset_path("emails.csv"))


def process_email(text: str) -> list[str]:
    """メール本文を、重複を除いた小文字の単語リストにする。

    原著は `list(set(text.lower().split()))`。**同じ単語が何度出ても 1 回** と
    数えるのがナイーブベイズのこの実装の前提である。
    """
    return list(set(text.lower().split()))


def train(emails: pd.DataFrame) -> NaiveBayesModel:
    """単語ごとに、スパム・ハムそれぞれの出現通数を数える。"""
    words: dict[str, WordCounts] = {}
    for _, email in emails.iterrows():
        for word in process_email(email["text"]):
            counts = words.setdefault(word, WordCounts())
            if email["spam"]:
                counts.spam += 1
            else:
                counts.ham += 1

    return NaiveBayesModel(
        words, Corpus(total=len(emails), spam=int(emails["spam"].sum()))
    )


def predict_bayes(model: NaiveBayesModel, word: str) -> float:
    """単語 1 つだけを見たときの、スパムである確率。

    ベイズの定理そのものではなく、単純に「その単語を含むメールのうち
    スパムの割合」を返す。原著もそう書いている。
    """
    counts = model.words[word.lower()]
    return counts.spam / (counts.spam + counts.ham)


def predict_naive_bayes(model: NaiveBayesModel, email: str) -> float:
    """メール全体を見たときの、スパムである確率。

    「単語の出現が互いに独立」と仮定して、単語ごとの尤度比を掛け合わせる。
    語彙にない単語は無視するので、知らない単語ばかりのメールは事前確率に落ちる。

    原著は最後に `np.compat.long(...)` を通しており、これは Python の `int` と
    同じで **小数点以下を切り捨てる**。切り捨てないと 8 桁目から数値が変わるので、
    そのまま写してある。`np.compat` は NumPy 2.0 で削除されたため、
    ここでは `int()` を直接使う。
    """
    corpus = model.corpus
    spams = [1.0]
    hams = [1.0]
    for word in set(email.lower().split()):
        counts = model.words.get(word)
        if counts is not None:
            spams.append(counts.spam / corpus.spam * corpus.total)
            hams.append(counts.ham / corpus.ham * corpus.total)

    product_spams = _product(spams) * corpus.spam
    product_hams = _product(hams) * corpus.ham
    # 原著の np.compat.long と同じ切り捨て
    truncated_spams = int(product_spams)
    truncated_hams = int(product_hams)
    return truncated_spams / (truncated_spams + truncated_hams)


def _product(values: list[float]) -> float:
    result = 1.0
    for value in values:
        result *= value
    return result
